File: src/main/java/no/nav/syfo/api/ressurser/azuread/VeilederAzureRessurs.java
package no.nav.syfo.api.ressurser.azuread;

import no.nav.security.oidc.context.OIDCRequestContextHolder;
import no.nav.security.spring.oidc.validation.api.ProtectedWithClaims;
import no.nav.syfo.api.domain.RSEnheter;
import no.nav.syfo.api.domain.RSVeilederInfo;
import no.nav.syfo.service.NorgService;
import no.nav.syfo.service.VeilederService;
import org.springframework.web.bind.annotation.*;

import javax.inject.Inject;

import static no.nav.syfo.api.mappers.RSEnhetMapper.enhet2rs;
import static no.nav.syfo.oidc.OIDCIssuer.AZURE;
import static no.nav.syfo.util.MapUtil.mapListe;
import static no.nav.syfo.util.OIDCUtil.getSubjectInternAzure;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@RestController
@RequestMapping(value = "/api/internad/veilederinfo")
@ProtectedWithClaims(issuer = AZURE)
public class VeilederAzureRessurs {

    private OIDCRequestContextHolder contextHolder;
    private NorgService norgService;
    private VeilederService veilederService;

    @Inject
    public VeilederAzureRessurs(
            OIDCRequestContextHolder contextHolder,
            NorgService norgService,
            VeilederService veilederService
    ) {
        this.contextHolder = contextHolder;
        this.norgService = norgService;
        this.veilederService = veilederService;
    }

    @GetMapping
    public RSVeilederInfo hentNavn() {
        return hentIdent(getSubjectInternAzure(contextHolder));
    }

    @GetMapping(produces = APPLICATION_JSON_VALUE)
    @RequestMapping(value = "/{ident}")
    public RSVeilederInfo hentIdent(@PathVariable("ident") String ident) {
        return new RSVeilederInfo()
                .navn(veilederService.hentVeileder(ident).navn)
                .ident(ident);
    }

    @GetMapping(value = "/enheter", produces = APPLICATION_JSON_VALUE)
    public RSEnheter hentEnheter() {
        return new RSEnheter()
                .enhetliste(mapListe(norgService.hentVeiledersNavEnheter(getSubjectInternAzure(contextHolder)), enhet2rs));
    }
}

File: src/main/kotlin/no/nav/syfo/controller/internad/veileder/VeilederAzureRessurs.kt
package no.nav.syfo.controller.internad.veileder

import no.nav.security.oidc.api.ProtectedWithClaims
import no.nav.security.oidc.context.OIDCRequestContextHolder
import no.nav.syfo.consumer.axsys.AxsysConsumer
import no.nav.syfo.consumer.axsys.AxsysEnhet
import no.nav.syfo.api.auth.OIDCIssuer
import no.nav.syfo.service.VeilederService
import no.nav.syfo.api.auth.OIDCUtil
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import java.util.stream.Collectors.toList
import javax.inject.Inject

@RestController
@RequestMapping(value = ["/api/internad/veilederinfo"])
@ProtectedWithClaims(issuer = OIDCIssuer.AZURE)
class VeilederAzureRessurs @Inject constructor(
        private val contextHolder: OIDCRequestContextHolder,
        private val axsysConsumer: AxsysConsumer,
        private val veilederService: VeilederService
) {
    @GetMapping
    fun hentNavn(): RSVeilederInfo {
        return hentIdent(OIDCUtil.getSubjectInternAzure(contextHolder))
    }

    @GetMapping(value = ["/{ident}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun hentIdent(@PathVariable("ident") ident: String): RSVeilederInfo {
        return RSVeilederInfo(
                navn = veilederService.hentVeilederNavn(ident).orElse("Fant ikke navn"),
                ident = ident
        )
    }

    @GetMapping(value = ["/enheter"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun hentEnheter(): RSEnheter {
        val axsysEnhetList = axsysConsumer.enheter(OIDCUtil.getSubjectInternAzure(contextHolder))
        return mapAxsysEnhet2RS(axsysEnhetList)
    }

    fun mapAxsysEnhet2RS(axsysEnhetList: List<AxsysEnhet>): RSEnheter {
        val enhetList = axsysEnhetList
                .stream()
                .map {
                    RSEnhet(
                            enhetId = it.enhetId,
                            navn = it.navn
                    )
                }
                .collect(toList())
        return RSEnheter(enhetliste = enhetList)
    }
}
